### ISSUE TITLE
Is there a a community space for this project?

### ISSUE THREAD

**Issue by ivan006** [mevdschee/php-crud-api#804]
This project from the onset seems very apt/intelligent. Are there any community space for users of this system to communicate and compare notes etc? I'd like to learn more and see the full potential.

**Comment by mevdschee** [comment]
Thank you for your kind words. 

>Are there any community space for users of this system to communicate and compare notes etc?

Not that I'm aware of (other than Github issues).

Kind regards, Maurits

**Comment by ivan006** [comment]
does this integrate with any other libraries like datatables?

**Comment by mevdschee** [comment]
>does this integrate with any other libraries like datatables?

Sure, why not?

**Comment by ivan006** [comment]
*scrutiny warning*

yeah but im wondering has anyone demonstrated how to do this so others can do it more easily? who uses this library and how have they used it? how has this library been useful to others? the concept sounds sensible but if its not implemented then me implementing it will take more effort. how can i get the most bang for my "buck" with this library?

**Comment by mevdschee** [comment]
Lot's of questions that I can't help you with (I don't know the answers). You may try it out and help other by publishing your findings.

If you need a user interface, then you may try: https://github.com/mevdschee/php-crud-ui

It is my take on a usable (and high performance) user interface for the API.

Kind regards, Maurits

**Comment by ivan006** [comment]
Cool do you mind demoing it in a youtube video? (screencastify chrome plugin can help record in case u dont know of any recording tools)

**Comment by mevdschee** [comment]
Why don't you try to run it with your own dataset? I'm here to help you out when you run into problems. Also, feel free to create a YouTube video about it.

**Comment by ivan006** [comment]
How about just a screenshot? Its a lot of hours for me to set up libraries

**Comment by ivan006** [comment]
Im very interested in crud generators and the like i have myself made a more dynamic crud generator that works for any table, this means you dont have to generate a new crud functionality for each table. My engine also allow for seeing records that are related to a record. E.g. here is record 2 in the commodity type table 

![image](https://user-images.githubusercontent.com/17248313/132343308-4dd68c4b-0d75-4ce3-82a0-58f4a6ddc65e.png)

**Comment by mevdschee** [comment]
![Screen Shot 2021-09-08 at 01 58 27](https://user-images.githubusercontent.com/1288217/132424726-ee01c201-4ff9-4793-9731-86d8b4a875a9.png)

---

![Screen Shot 2021-09-08 at 01 58 40](https://user-images.githubusercontent.com/1288217/132424725-1681c0ae-fbf1-4096-ab3f-32e527a8e054.png)

----

![Screen Shot 2021-09-08 at 01 58 46](https://user-images.githubusercontent.com/1288217/132424724-1d9bdaed-e707-4879-bb63-7c25a2a22b16.png)

---

![Screen Shot 2021-09-08 at 01 58 54](https://user-images.githubusercontent.com/1288217/132424720-7a7ede84-75b7-4e48-ac02-a12ff51ddd4e.png)

**Comment by mevdschee** [comment]
>Im very interested in crud generators and the like i have myself made a more dynamic crud generator that works for any table

Nice, did you publish it? Can I try it out?

**Comment by ivan006** [comment]
Yes but do u mean as an installable library? If so no its only available as a full app [https://github.com/ivan006/Bg---table-page-3---free-php-spreadsheet-database-hybrid](https://github.com/ivan006/Bg---table-page-3---free-php-spreadsheet-database-hybrid) I will make a youtube video on how to set it up when i have the time thanks for ur interest.
And thanks for the screenshots.

So in your app i see you can  "view" a record so when u are viewing a record can u see related records? boths parents and children?

**Comment by mevdschee** [comment]
>when u are viewing a record can u see related records? boths parents and children?

You can only see a link to the filtered list (of children).

![Screen Shot 2021-09-08 at 15 29 34](https://user-images.githubusercontent.com/1288217/132518583-92cdd17c-578e-4c2e-ad55-8c48f35277be.png)

**Comment by ivan006** [comment]
Wow and thats "post='blog started'" thats cool. I understand "post" refers to the table but what does "blog started" refer to is searching all columns or a specific column if its a specific column then how do u flag which column is priority?

**Comment by mevdschee** [comment]
'Post' is short for 'post_id' (a fk filter on the comments table) and 'blog started' is the default representation of a record in the related table (the value in the 'name' field)

**Comment by ivan006** [comment]
Ok cool, and then this unfortunately contrains the developer to have to create a column with the name "name". Yeah my system also has those kinds of contraints for instant the primary key has to called "id" which makes the system a bit rigid, oh well u cant win them all.

**Comment by mevdschee** [comment]
> this unfortunately contains the developer to have to create a column with the name "name".

It returns the first text column, in this case 'name', see:

    private function getDisplayColumn(string $table, string $action)
    {
        $properties = $this->getProperties($table, $action);

        foreach ($properties as $field => $property) {
            if ($property['type'] == 'string') {
                return $field;
            }
        }
        return false;
    }

I hope this helps you in any way. Kind regards, Maurits

**Comment by mevdschee** [comment]
>Yeah my system also has those kinds of constraints for instance the primary key has to called "id" which makes the system a bit rigid

I don't have that constraint. I request the primary key from the information schema (it may not be composite), see:

    private function getTablePrimaryKeysSQL(): string
    {
        switch ($this->driver) {
            case 'mysql':
                return 'SELECT "COLUMN_NAME" FROM "INFORMATION_SCHEMA"."KEY_COLUMN_USAGE" WHERE "CONSTRAINT_NAME" = \'PRIMARY\' AND "TABLE_NAME" = ? AND "TABLE_SCHEMA" = ?';
            case 'pgsql':
                return 'SELECT a.attname AS "COLUMN_NAME" FROM pg_attribute a JOIN pg_constraint c ON (c.conrelid, c.conkey[1]) = (a.attrelid, a.attnum) JOIN pg_class pgc ON pgc.oid = a.attrelid WHERE pgc.relname = ? AND \'\' <> ? AND c.contype = \'p\'';
            case 'sqlsrv':
                return 'SELECT c.NAME as "COLUMN_NAME" FROM sys.key_constraints kc inner join sys.objects t on t.object_id = kc.parent_object_id INNER JOIN sys.index_columns ic ON kc.parent_object_id = ic.object_id and kc.unique_index_id = ic.index_id INNER JOIN sys.columns c ON ic.object_id = c.object_id AND ic.column_id = c.column_id WHERE kc.type = \'PK\' and t.object_id = OBJECT_ID(?) and \'\' <> ?';
            case 'sqlite':
                return 'SELECT "name" as "COLUMN_NAME" FROM pragma_table_info(?) WHERE "pk"=1 AND \'\' <> ?';
        }
    }

from: https://github.com/mevdschee/php-crud-api/blob/main/src/Tqdev/PhpCrudApi/Database/GenericReflection.php#L66

I hope this helps you in any way. Kind regards, Maurits

**Comment by mevdschee** [comment]
>my system also has those kinds of constraints

My constraints are:

  - Primary keys should either be auto-increment (from 1 to 2^53) or UUID
  - Composite primary and composite foreign keys are not supported
  - Complex writes (transactions) are not supported
  - Complex queries calling functions (like "concat" or "sum") are not supported
  - Database must support and define foreign key constraints
  - SQLite cannot have bigint typed auto incrementing primary keys
  - SQLite does not support altering table columns (structure)

from: https://github.com/mevdschee/php-crud-api#limitations

**Comment by ivan006** [comment]
> 
> 
> > Im very interested in crud generators and the like i have myself made a more dynamic crud generator that works for any table
> 
> Nice, did you publish it? Can I try it out?

As promised here it finally is https://www.youtube.com/playlist?list=PLDjAbnn_RxMIwOTO-3ciB1E1UmB8MEv4u

**Comment by mevdschee** [comment]
Wow.. what a work.. I mean.. the video.. the code.. everything! Congrats and thank you for sharing.

**Comment by ivan006** [comment]
> Wow.. what a work.. I mean.. the video.. the code.. everything! Congrats and thank you for sharing.

So kind thank you

**Comment by ivan006** [comment]
Let me know when/if u get round to installing my app or just watching the videos
